mlir/include/mlir/Linker/LLVMLinkerMixin.h

```diff
@@ -340,6 +340,48 @@ class LLVMLinkerMixin {
       return ConflictResolution::LinkFromSrc;
     }
 
+    // Handle external weak linkage conflicts
+    if (isExternalWeakLinkage(srcLinkage)) {
+      if (isExternalWeakLinkage(dstLinkage)) {
+        // Both external weak, keep dst
+        return ConflictResolution::LinkFromDst;
+      }
+      if (isExternalLinkage(dstLinkage) || isWeakForLinker(dstLinkage)) {
+        // External weak loses to external or other weak
+        return ConflictResolution::LinkFromDst;
+      }
+      // External weak wins over declarations
+      if (dstIsDeclaration) {
+        return ConflictResolution::LinkFromSrc;
+      }
+      return ConflictResolution::LinkFromDst;
+    }
+
+    if (isExternalWeakLinkage(dstLinkage)) {
+      if (isExternalLinkage(srcLinkage) || isWeakForLinker(srcLinkage)) {
+        // External or weak src wins over external weak dst
+        return ConflictResolution::LinkFromSrc;
+      }
+      // External weak wins over declarations
+      if (srcIsDeclaration) {
+        return ConflictResolution::LinkFromDst;
+      }
+      return ConflictResolution::LinkFromSrc;
+    }
+
+    // Handle available externally + external definition conflicts
+    if (isAvailableExternallyLinkage(srcLinkage) &&
+        isExternalLinkage(dstLinkage)) {
+      // External definition wins over available externally
+      return ConflictResolution::LinkFromDst;
+    }
+
+    if (isAvailableExternallyLinkage(dstLinkage) &&
+        isExternalLinkage(srcLinkage)) {
+      // External definition wins over available externally
+      return ConflictResolution::LinkFromSrc;
+    }
```

Comment:
This part seems a bit strange to me…
`llvm-link` does not have to handle this (see [their implementation](https://github.com/llvm/llvm-project/blob/ad5778ff161110c04323d758c1c25a6d7640b10b/llvm/lib/Linker/LinkModules.cpp#L240) of conflict resolution) and it seems like quite a lot of potentially unresolved conflicts they would encounter.

Was this maybe necessary to add because of some other bug? Where/when did you encounter these unresolved conflicts?

Reply:
I see. I will try to give it a look to see if that's a symptom of some other bug or if this is a legal case.
IIUIC, there are some invalid cases that should not pop-up (or should not allow linking to continue), so that's why I'm trying to be careful around it :)